lib/ngx/re.lua

```diff
@@ -0,0 +1,163 @@
+-- I hereby assign copyright in this code to the lua-resty-core project,
+-- to be licensed under the same terms as the rest of the code.
+
+local ffi = require 'ffi'
+local bit = require "bit"
+local base = require "resty.core.base"
+local core_regex = require "resty.core.regex"
```

Comment:
I'm not feeling very comfortable with loading the whole `resty.core.regex` module just for `ngx.re`. Since requiring `resty.core.regex` has the side effect of overriding most of the standard `ngx.re` API functions.

I think a better way is to put the main implementations of `resty.core.regex` here, in `ngx.re`, and export functions for `resty.core.regex` to use.

I understand that it is a big change for this PR, maybe we can do this refactoring in the future.